File: Sdks/AdminApi.Sdk/src/Endpoints/Clients/ClientsEndpoint.cs
using Backbone.AdminApi.Sdk.Endpoints.Clients.Types;
using Backbone.AdminApi.Sdk.Endpoints.Clients.Types.Requests;
using Backbone.AdminApi.Sdk.Endpoints.Clients.Types.Responses;
using Backbone.BuildingBlocks.SDK.Endpoints.Common;
using Backbone.BuildingBlocks.SDK.Endpoints.Common.Types;

namespace Backbone.AdminApi.Sdk.Endpoints.Clients;

public class ClientsEndpoint(EndpointClient client) : AdminApiEndpoint(client)
{
    public async Task<ApiResponse<ListClientsResponse>> ListClients()
    {
        return await _client.Get<ListClientsResponse>($"api/{API_VERSION}/Clients");
    }

    public async Task<ApiResponse<ClientInfo>> GetClient(string id)
    {
        return await _client.Get<ClientInfo>($"api/{API_VERSION}/Clients/{id}");
    }

    public async Task<ApiResponse<CreateClientResponse>> CreateClient(CreateClientRequest request)
    {
        return await _client.Post<CreateClientResponse>($"api/{API_VERSION}/Clients", request);
    }

    public async Task<ApiResponse<ChangeClientSecretResponse>> ChangeClientSecret(string id, ChangeClientSecretRequest request)
    {
        return await _client.Patch<ChangeClientSecretResponse>($"api/{API_VERSION}/Clients/{id}/ChangeSecret", request);
    }

    public async Task<ApiResponse<UpdateClientResponse>> UpdateClient(string id, UpdateClientRequest request)
    {
        return await _client.Put<UpdateClientResponse>($"api/{API_VERSION}/Clients/{id}", request);
    }

    public async Task<ApiResponse<EmptyResponse>> DeleteClient(string id)
    {
        return await _client.Delete<EmptyResponse>($"api/{API_VERSION}/Clients/{id}");
    }
}

File: Applications/AdminUi/packages/admin_api_sdk/lib/src/endpoints/clients_endpoint.dart
import 'package:admin_api_types/admin_api_types.dart';

import '../types/types.dart';
import 'endpoint.dart';

class ClientsEndpoint extends Endpoint {
  ClientsEndpoint(super._dio);

  Future<ApiResponse<List<ClientOverview>>> getClients() =>
      get('/api/v1/Clients', transformer: (e) => (e as List).map(ClientOverview.fromJson).toList());

  Future<ApiResponse<CreateClientResponse>> createClient({
    required String defaultTier,
    String? clientId,
    String? clientSecret,
    String? displayName,
    int? maxIdentities,
  }) => post(
    '/api/v1/Clients',
    data: {
      'defaultTier': defaultTier,
      'clientId': clientId,
      'clientSecret': clientSecret,
      'displayName': displayName,
      'maxIdentities': maxIdentities,
    },
    transformer: CreateClientResponse.fromJson,
  );

  Future<ApiResponse<Client>> getClient(String clientId) => get('/api/v1/Clients/$clientId', transformer: Client.fromJson);

  Future<ApiResponse<ChangeClientSecretResponse>> changeClientSecret(String clientId, {required String? newSecret}) =>
      patch('/api/v1/Clients/$clientId/ChangeSecret', data: {'newSecret': newSecret}, transformer: ChangeClientSecretResponse.fromJson);

  Future<ApiResponse<UpdateClientResponse>> updateClient(String clientId, {required String defaultTier, required int? maxIdentities}) => put(
    '/api/v1/Clients/$clientId',
    data: {'defaultTier': defaultTier, 'maxIdentities': maxIdentities},
    transformer: UpdateClientResponse.fromJson,
  );

  Future<ApiResponse<void>> deleteClient(String clientId) =>
      delete('/api/v1/Clients/$clientId', expectedStatus: 204, transformer: (e) {}, allowEmptyResponse: true);
}
